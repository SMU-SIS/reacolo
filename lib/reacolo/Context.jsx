import React from 'react';

const Context = () => {
  throw new Error('Context components cannot be rendered directly. They must be contained either ' +
    'in a Filter component or a parent Context component.');
};

Context.propTypes = {
  default: React.PropTypes.bool,
  children: React.PropTypes.node
};

Context.defaultProps = {
  default: false
};

export default Context;
