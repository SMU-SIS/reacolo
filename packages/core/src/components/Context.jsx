import React from 'react';
// import omit from 'object.omit';
import PropTypes from 'prop-types';
import matchContext from '../filtering/match-context.js';
import createMatcher from '../filtering/create-matcher.js';
import connect from '../connect/connect.jsx';

/**
 * Context component. Renders only if its value matches its match properties.
 * @param {object} props The component's properties.
 * @return {Element} A react element.
 */
export const Context = props => {
  const matcher = createMatcher(props);
  const shouldRender = matchContext(props.contextValue, matcher);
  if (!shouldRender) return null;
  if (props.render) return props.render();
  if (props.component) return <props.component />;
  if (props.children) return props.children;
  throw new Error(
    'No render function nor component properties have been provided to <Context>',
  );
};

Context.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  contextValue: PropTypes.object.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  default: PropTypes.bool,
  render: PropTypes.func,
  component: PropTypes.func,
  children: PropTypes.element,
};

Context.defaultProps = {
  default: false,
  render: undefined,
  component: undefined,
  children: undefined,
};

// Default export is a <ConnectedContext>, a <Context> that fetches its value
// automatically from the React context.
export default connect(undefined, undefined, context => ({
  contextValue: context,
}))(Context);
