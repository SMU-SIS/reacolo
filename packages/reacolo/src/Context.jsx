import bindContext from './bind-context';

export default bindContext();
