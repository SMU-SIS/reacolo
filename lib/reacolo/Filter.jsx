import React from 'react';
import except from 'except';
import Context from './Context';
import wouldPass from './filtering/would-pass';

const OMITTED_CONTEXT_PROPS_FROM_TARGETS = ['default', 'children'];
const getTargetsFromProps = props => except(props, OMITTED_CONTEXT_PROPS_FROM_TARGETS);

const lookupFilters = (children, context) => {
  // Look for the first filters that passes.
  const passingChild = children
    .find(child => wouldPass(context, getTargetsFromProps(child.props)));
  // If no filter passed, look for a default filter.
  const selectedChild = passingChild || children.find(child => child.props.default);

  // If nothing have been selected, return.
  if (!selectedChild) return undefined;

  const selectionsChildren = selectedChild.props.children;

  // Else, look at the first children to check if there is nested filters.
  const selectionsChildrenArray = React.Children.toArray(selectionsChildren);

  if (selectionsChildrenArray[0]) {
    if (selectionsChildrenArray[0].type === Context) {
      // And if it is the case, recursively call selectFilter on these children.
      return lookupFilters(selectionsChildrenArray, context);
    }
    return React.Children.only(selectionsChildren);
  }
  return undefined;
};

const Filter = ({ children, context }) => lookupFilters(React.Children.toArray(children), context)
  || null;

export default Filter;
