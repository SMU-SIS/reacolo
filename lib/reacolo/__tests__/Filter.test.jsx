import React from 'react';
import renderer from 'react-test-renderer';
import Filter from '../Filter';
import Context from '../Context';
import wouldPass from '../filtering/would-pass';

jest.mock('../filtering/would-pass');

beforeEach(() => {
  wouldPass.mockReset();
  wouldPass.mockImplementation(() => false);
});

describe('Filter with non nested Context', () => {
  it('properly calls wouldPass until it returns true', () => {
    wouldPass.mockReturnValueOnce(false)
             .mockReturnValueOnce(true);
    renderer.create(
      <Filter context={{ contextProp: 'bar' }}>
        <Context prop1="val1" default><div>child1</div></Context>
        <Context prop2="val2"><div>child2</div></Context>
        <Context prop3="val3"><div>child3</div></Context>
      </Filter>
    );
    expect(wouldPass.mock.calls).toEqual([
      [
        { contextProp: 'bar' },
        { prop1: 'val1' }
      ], [
        { contextProp: 'bar' },
        { prop2: 'val2' }
      ]
    ]);
  });
  it('renders the first context that passes', () => {
    wouldPass.mockImplementation((_, { pass }) => pass);
    const component = renderer.create(
      <Filter context={{ roles: {} }}>
        <Context><div>Not to be rendered</div></Context>
        <Context pass><div>To be rendered</div></Context>
        <Context pass><div>Not to be rendered</div></Context>
      </Filter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('does not render if no contexts would pass and there is no default context', () => {
    wouldPass.mockReturnValue(false);
    const component = renderer.create(
      <Filter context={{ roles: {} }}>
        <Context><div>Not to be rendered</div></Context>
        <Context><div>Not to be rendered</div></Context>
      </Filter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the default context if no contexts would pass', () => {
    wouldPass.mockReturnValue(false);
    const component = renderer.create(
      <Filter context={{ roles: {} }}>
        <Context><div>Not to be rendered</div></Context>
        <Context default><div>Default Filter Content</div></Context>
        <Context><div>Not to be rendered</div></Context>
      </Filter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the context that passes even if there is a default context', () => {
    wouldPass.mockImplementation((_, { pass }) => pass);
    const component = renderer.create(
      <Filter context={{ roles: {} }}>
        <Context><div>Not to be rendered</div></Context>
        <Context pass><div>To be rendered</div></Context>
        <Context default><div>Default Filter Content</div></Context>
        <Context pass><div>Not to be rendered</div></Context>
        <Context><div>Not to be rendered</div></Context>
      </Filter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the context that passes even if it comes after the default context', () => {
    wouldPass.mockImplementation((_, { pass }) => pass);
    const component = renderer.create(
      <Filter context={{ roles: {} }}>
        <Context><div>Not to be rendered</div></Context>
        <Context default><div>Default Filter Content</div></Context>
        <Context pass><div>To be rendered</div></Context>
        <Context pass><div>Not to be rendered</div></Context>
      </Filter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('Filter with nested Context components', () => {
  let component;
  beforeEach(() => {
    wouldPass.mockImplementation((_, { pass }) => pass);
    component = renderer.create(
      <Filter context={{ contextProp: 'foo' }}>
        <Context prop1="val1" default>
          <Context prop11="val11"><div>child11</div></Context>
          <Context prop12="val12" pass><div>child12</div></Context>
        </Context>
        <Context prop2="val2" pass>
          <Context prop21="val21"><div>child21</div></Context>
          <Context prop22="val22" pass>
            <Context prop221="val221"><div>child221</div></Context>
            <Context prop222="val222" default><div>child222 (to be rendered)</div></Context>
            <Context prop223="val223"><div>child223</div></Context>
          </Context>
          <Context prop23="val23" pass><div>child23</div></Context>
        </Context>
        <Context prop3="val3"><div>child3</div></Context>
      </Filter>
    );
  });
  it('properly calls wouldPass on nested contexts if their parent passed', () => {
    expect(wouldPass.mock.calls).toEqual([
      [
        { contextProp: 'foo' },
        { prop1: 'val1' }
      ], [
        { contextProp: 'foo' },
        { prop2: 'val2', pass: true }
      ], [
        { contextProp: 'foo' },
        { prop21: 'val21' }
      ], [
        { contextProp: 'foo' },
        { prop22: 'val22', pass: true }
      ], [
        { contextProp: 'foo' },
        { prop221: 'val221' }
      ], [
        { contextProp: 'foo' },
        { prop222: 'val222' }
      ], [
        { contextProp: 'foo' },
        { prop223: 'val223' }
      ]
    ]);
  });
  it('properly renders the right nested context', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
