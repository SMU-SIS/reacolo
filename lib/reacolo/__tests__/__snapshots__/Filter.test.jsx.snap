// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter with nested Context components properly renders the right nested context 1`] = `
<div>
  child222 (to be rendered)
</div>
`;

exports[`Filter with non nested Context does not render if no contexts would pass and there is no default context 1`] = `null`;

exports[`Filter with non nested Context renders the context that passes even if it comes after the default context 1`] = `
<div>
  To be rendered
</div>
`;

exports[`Filter with non nested Context renders the context that passes even if there is a default context 1`] = `
<div>
  To be rendered
</div>
`;

exports[`Filter with non nested Context renders the default context if no contexts would pass 1`] = `
<div>
  Default Filter Content
</div>
`;

exports[`Filter with non nested Context renders the first context that passes 1`] = `
<div>
  To be rendered
</div>
`;
